### PR TITLE
[BPF] BPFExcludeCIDRsFromNAT allows excluding service IPs from being NATed

### PR DESCRIPTION
**Pull request by tomastigera**
The main use case is node-local dns cache. In iptables mode with k8s provided kube-proxy, the cache installs such iptables rules that snatch the traffic before the kube-proxy rules, which would do the NAT.

However, in ebpf NAT is done way before packets may hit iptables. Often they do not hit iptables at all. Therefore, we need to tell our kube-proxy, not to translate a service IP that we want to let dns cache to process.

Unfortunatelly, we cannot annotate the service as the annotations are not propagate to the kube-proxy code by the k8s front-end.

To setup a node-local dns cache, set BPFExcludeCIDRsFromNAT as follows:

kubedns=`kubectl get svc kube-dns -n kube-system -o jsonpath={.spec.clusterIP}`

BPFExcludeCIDRsFromNAT=$kubedns

fixes https://github.com/projectcalico/calico/issues/4689

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
ebpf: setting BPFExcludeIPsFromNAT allows node-local dns cache to work
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/dataplane/linux/bpf_ep_mgr.go
Comment (fasaxc) [inline felix/dataplane/linux/bpf_ep_mgr.go:519]: What about the reverse, IPv4 CIDR and IPv6 support enabled?  
```suggestion
			if (cidr.Version() == 6) != m.ipv6Enabled {
```
do you need the cast?

File: felix/bpf-gpl/tc.c
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:414]: Do we need to mark the conntrack entry too?  Otherwise only the first packet will skip the FIB.

**Review by fasaxc — CHANGES_REQUESTED** [review]

File: api/pkg/apis/projectcalico/v3/felixconfig.go
Comment (tomastigera) [reply api/pkg/apis/projectcalico/v3/felixconfig.go:555]: yeah agree, noticed it myself too :sweat_smile:

File: felix/dataplane/linux/bpf_ep_mgr.go
Comment (tomastigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:519]: Atm if `m.ipv6Enabled` then ipv4 should not be considered. As part of the dual stack effort, this will need to be addressed, but idk how that is going to pan out yet.

File: felix/bpf-gpl/tc.c
Comment (tomastigera) [reply felix/bpf-gpl/tc.c:414]: Correct, that is done where we create the ct entry here. That is an already established process. https://github.com/tomastigera/project-calico-calico/blob/tomas-bpf-node-local-dns/felix/bpf-gpl/tc.c#L1231-L1232

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Review by fasaxc — APPROVED** [review]

**Pull request merged**
